In [32]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn import linear_model as lm

In [33]:
df=pd.read_csv('https://raw.githubusercontent.com/xiaowanzisj/Data-Science-1/refs/heads/main/dataset_lm.csv')

In [34]:
df.head()

Dependent Var  Explanatory Var #1  Explanatory Var #2  Explanatory Var #3  \
0      56.293458           13.698667           50.639873                   0   
1      58.473431            2.714725           65.845845                   1   
2      94.195330           11.618072           65.072497                   0   
3      29.074583            0.818623           45.408996                   1   
4      86.035569            9.077544           73.548021                   0   

   Explanatory Var #4  Explanatory Var #5  Explanatory Var #6  \
0          -18.568035           45.121911           11.412501   
1          -25.105932           47.190213           10.080280   
2           -7.897464           52.163036           11.057301   
3          -18.316132           54.356714            5.029029   
4          -19.204165           47.186807           12.128134   

   Explanatory Var #7  Explanatory Var #8  Explanatory Var #9  \
0           56.410757                   2          -12.281132   
1           65.383107                   3          -36.763585   
2           82.812717                   0          -15.733547   
3           48.812471                   1          -12.825591   
4           62.520911                   2          -13.804860   

   Explanatory Var #10  Explanatory Var #11  Explanatory Var #12  \
0            38.996909            -3.010548            49.195073   
1            51.654939             4.991111            45.591729   
2            48.913837            -2.457696            56.608806   
3            45.851732            14.974177            47.362594   
4            47.765904             9.593982            53.700562   

   Explanatory Var #13  Explanatory Var #14  Explanatory Var #15  
0                    0           -21.153143            46.919314  
1                    0            -6.474403            53.383508  
2                    0           -27.903299            48.515026  
3                    1           -10.064411            55.266254  
4                    0           -17.546302            48.150543

In [35]:
df.describe()

Dependent Var  Explanatory Var #1  Explanatory Var #2  \
count     422.000000          422.000000          422.000000   
mean       67.838396            7.762058           59.725134   
std        25.676960            7.013243            9.579112   
min        -2.997183           -4.832834           44.124858   
25%        49.482037            1.720182           51.617692   
50%        66.780110            7.905455           59.735139   
75%        86.801496           13.684104           67.870073   
max       133.384795           19.973331           76.973576   

       Explanatory Var #3  Explanatory Var #4  Explanatory Var #5  \
count          422.000000          422.000000          422.000000   
mean             0.500000          -16.078411           50.133715   
std              0.500593            8.111197            6.975674   
min              0.000000          -29.774797           30.009511   
25%              0.000000          -23.792637           45.423422   
50%              0.500000          -15.875481           50.093602   
75%              1.000000           -9.185191           54.896583   
max              1.000000           -2.060708           70.365951   

       Explanatory Var #6  Explanatory Var #7  Explanatory Var #8  \
count          422.000000          422.000000          422.000000   
mean            10.420808           61.887309            1.566351   
std             11.520725           14.819969            1.115260   
min             -9.828552           34.093154            0.000000   
25%             -0.174835           49.952772            1.000000   
50%             10.422513           62.554591            2.000000   
75%             21.059713           74.441216            3.000000   
max             29.994610           86.895006            3.000000   

       Explanatory Var #9  Explanatory Var #10  Explanatory Var #11  \
count          422.000000           422.000000           422.000000   
mean           -24.511691            49.728665             7.740874   
std              8.500500             6.910690             7.102714   
min            -39.979696            26.436407            -4.949728   
25%            -32.037707            45.283603             1.838210   
50%            -23.767548            49.842746             8.055297   
75%            -17.419390            54.576381            14.020396   
max            -10.129522            68.201681            19.992891   

       Explanatory Var #12  Explanatory Var #13  Explanatory Var #14  \
count           422.000000           422.000000           422.000000   
mean             60.765794             0.500000           -16.673762   
std               9.525835             0.500593             8.106466   
min              44.158200             0.000000           -29.750628   
25%              52.849792             0.000000           -23.797763   
50%              60.773906             0.500000           -17.373619   
75%              69.262757             1.000000            -9.627544   
max              76.639179             1.000000            -2.003168   

       Explanatory Var #15  
count           422.000000  
mean             50.066901  
std               6.794584  
min              32.118882  
25%              45.825931  
50%              50.038667  
75%              54.962602  
max              69.147818

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

In [37]:
df.shape

(422, 16)

# Part C

In [38]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso

In [39]:
# Split the dataset into two as the training and test sets
y = df['Dependent Var']
X = df.drop('Dependent Var', axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

In [41]:
# Run the Lasso model with alpha=1 and estimate the coefficients using the training set
model_lasso = lm.Lasso(alpha=1).fit(X_train,y_train)

In [44]:
model_lasso.coef_

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

Lasso has set some coefficients to zero, indicating that these variables may not contribute significantly to the model. 

In [45]:
# Calculate the mean absolute percentage error using the test set
from sklearn.metrics import mean_absolute_percentage_error

# Predict on the test set using the fitted Lasso model
y_pred = model_lasso.predict(X_test)

# Calculate the Mean Absolute Percentage Error 
mape = mean_absolute_percentage_error(y_test, y_pred)
mape


np.float64(0.04432190198291586)

1. The Mean Absolute Percentage Error (MAPE), which indicates the average percentage error between the predicted and actual values in the test set for the Lasso model on the test set, is approximately 4.43%. 
2. A 4.43% MAPE number seems reasonable, only depending on the context and the specific field or industry where the data comes from. 


In [48]:
# Try alpha=10 first to check how the MAPE value changes 
model_lasso_10 = lm.Lasso(alpha=10).fit(X_train,y_train)
model_lasso_10.coef_

array([ 1.07859582,  1.58170805,  0.        ,  1.93984057, -0.73184967,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.        ,  0.        , -0.        , -0.        , -0.        ])

In [50]:
y_pred = model_lasso_10.predict(X_test)
mape = mean_absolute_percentage_error(y_test, y_pred)
mape

np.float64(0.08587094351590434)

The MAPE value increases significantly to 8.59% 

In [52]:
# Apply Ridge Regression to our data
from sklearn.linear_model import Ridge
model_ridge = lm.Ridge(alpha=1).fit(X_train,y_train)
model_ridge.coef_

array([ 1.29954993e+00,  1.69972328e+00,  6.07547327e+00,  2.09998671e+00,
       -9.01138066e-01, -2.76774698e-04,  1.39022392e-04, -1.29553546e-03,
        1.28574039e-04, -1.91998438e-04,  9.64691394e-04,  1.51252792e-04,
       -7.24514808e-03,  4.88963422e-05, -1.75560566e-03])

In [53]:
# Predict on the test set using the fitted Ridge model
y_pred_ridge = model_ridge.predict(X_test)

In [54]:
# Calculate the MAPE for the Ridge model
mape_ridge = mean_absolute_percentage_error(y_test, y_pred_ridge)
mape_ridge

np.float64(0.0013183099780280923)

the MAPE decreased significantly to 0.13%

In [55]:
# Use LassoCV to find the optimal alpha to minimize MAPE
from sklearn.linear_model import LassoCV

# Setting cv=5 for cross-validation
model_lasso_cv = LassoCV(alphas=None, cv=5, random_state=42)
model_lasso_cv.fit(X_train, y_train)

# Predict on the test set with the optimal alpha found by cross-validation
y_pred_cv = model_lasso_cv.predict(X_test)

# Calculate the MAPE with this optimal alpha
optimal_alpha = model_lasso_cv.alpha_
mape_cv = mean_absolute_percentage_error(y_test, y_pred_cv)
optimal_alpha, mape_cv



(np.float64(0.17143335132581347), np.float64(0.007598538060298686))

1. The results from LassoCV mean that when alpha=0.171, the MAPE is 0.76%，improved a lot from 4.43% when alpha was set at 1. 
2. However, the minimal MAPE from Lasso model which is around 0.76% is still higher than that from Ridge Regression model which is 0.13%. It suggests Ridge provides more accurate predictions on this dataset. This could be due to Ridge Regression’s ability to handle multicollinearity without setting coefficients to zero, effectively utilizing all features for a smoother fit. 